File: http-core/src/main/java/org/apache/pekko/http/javadsl/model/TransferEncoding.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * license agreements; and to You under the Apache License, version 2.0:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * This file is part of the Apache Pekko project, which was derived from Akka.
 */

/*
 * Copyright (C) 2009-2022 Lightbend Inc. <https://www.lightbend.com>
 */

package org.apache.pekko.http.javadsl.model;

import org.apache.pekko.http.impl.util.Util;
import org.apache.pekko.http.javadsl.model.headers.EntityTagRanges;

import java.util.Map;

/** @see TransferEncodings for convenience access to often used values. */
public abstract class TransferEncoding {
  public abstract String name();

  public abstract Map<String, String> getParams();

  public static TransferEncoding createExtension(String name) {
    return new org.apache.pekko.http.scaladsl.model.TransferEncodings.Extension(
        name, Util.emptyMap);
  }

  public static TransferEncoding createExtension(String name, Map<String, String> params) {
    return new org.apache.pekko.http.scaladsl.model.TransferEncodings.Extension(
        name, Util.convertMapToScala(params));
  }
}

File: http-core/src/main/scala/org/apache/pekko/http/scaladsl/model/TransferEncoding.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * license agreements; and to You under the Apache License, version 2.0:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * This file is part of the Apache Pekko project, which was derived from Akka.
 */

/*
 * Copyright (C) 2009-2022 Lightbend Inc. <https://www.lightbend.com>
 */

package org.apache.pekko.http.scaladsl.model

import org.apache.pekko
import pekko.http.impl.util.{ Renderable, Rendering, SingletonValueRenderable }
import pekko.http.javadsl.{ model => jm }
import pekko.http.impl.util.JavaMapping.Implicits._

sealed abstract class TransferEncoding extends jm.TransferEncoding with Renderable {
  def name: String
  def params: Map[String, String]

  def getParams: java.util.Map[String, String] = params.asJava
}

object TransferEncodings {
  protected abstract class Predefined extends TransferEncoding with SingletonValueRenderable {
    def name: String = value
    def params: Map[String, String] = Map.empty
  }

  case object chunked extends Predefined
  case object compress extends Predefined
  case object deflate extends Predefined
  case object gzip extends Predefined
  case object trailers extends Predefined
  final case class Extension(name: String, params: Map[String, String] = Map.empty) extends TransferEncoding {
    def render[R <: Rendering](r: R): r.type = {
      r ~~ name
      params.foreach { case (k, v) => r ~~ "; " ~~ k ~~ '=' ~~# v }
      r
    }
  }
}
